src/EditorFeatures/Core/Implementation/PasteTracking/PasteTrackingService.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.PasteTracking
+{
+    [Export(typeof(IPasteTrackingService)), Shared]
+    internal class PasteTrackingService : IPasteTrackingService
+    {
+        private readonly IThreadingContext _threadingContext;
+        private readonly ITextBufferAssociatedViewService _textBufferAssociatedViewService;
+
+        [ImportingConstructor]
+        public PasteTrackingService(IThreadingContext threadingContext, ITextBufferAssociatedViewService textBufferAssociatedViewService)
+        {
+            _threadingContext = threadingContext;
+            _textBufferAssociatedViewService = textBufferAssociatedViewService;
+        }
+
+        public bool TryGetPastedTextSpan(Document document, out TextSpan textSpan)
+        {
+            if (!TryGetTextBuffer(document, out var textBuffer))
+            {
+                textSpan = default;
+                return false;
+            }
+
+            // `PropertiesCollection` is thread-safe
+            return textBuffer.Properties.TryGetProperty(this, out textSpan);
+        }
+
+        internal void RegisterPastedTextSpan(Document document, TextSpan textSpan)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            if (!TryGetTextBuffer(document, out var textBuffer))
+            {
+                return;
+            }
+
+            var textView = _textBufferAssociatedViewService
+                .GetAssociatedTextViews(textBuffer)
+                .FirstOrDefault(view => view.HasAggregateFocus);
+
+            if (textView is null)
+            {
+                return;
+            }
+
+            textView.Closed += ClearTracking;
```

Comment:
what happen if paste happened from 2 different view/buffer? will closing the first view remove tracking for the second one?

Reply:
I see, so 2 buffer case is fine. but still 2 views over 1 buffer case. closing 1 view seems remove tracking even though buffer didn't change.

Reply:
I think for 2 view case, we do some ref counting. and have some helper method. for for buffer tagger or something like that. you can do that or might be okay to ignroe since it might be just minor corner case noboy care.

Reply:
I feel like tracking open views would add a lot of complexity to be correct in this corner case.

Reply:
the critical thing to me is: by not tracking views:
1. does this just mean that in some esoteric case we'll miss like a single paste?
2. does this mean we can end up in a state where paste tracking now longer works for a document?

Reply:
The former is ok.  The latter would not be.